### PR TITLE
Cmake dep

### PR DESCRIPTION
**Pull request by wdobbe**
- Added code in cmake config files to make cross-compile work.
- Updated readme with cross-compile information.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -830,6 +756,73 @@ JSON printer and parser can be generated using the --json flag or
 some certain runtime library compile time flags that can optimize out
 printing symbolic enums, but these can also be disabled at runtime.
 
+## Using flatcc in your CMakeLists.txt
```
Comment (mikkelfj) [inline README.md]:
```suggestion
## Using flatcc in CMake projects```

File: README.md [hunk 2]
```diff
@@ -830,6 +756,73 @@ JSON printer and parser can be generated using the --json flag or
...
+        OUTPUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/datadef"
+        BUILDER
+        VERIFIER
+    )
+
+    add_executable(my_main main.c)
+    target_link_libraries(seclif_protocol PRIVATE flatcc::runtime flatcc_generated::seclif_protocol)
+
+Cross building with flatbuffers on cmake is possible by providing the install location of the build architecture flatcc package via environment variable flatccCli_ROOT or via cmake variable flatccCli_ROOT. The flatcc_generate_sources function will then use the *build* architecture flatcc cli executable to generate the header files and link your target to the *host* architecture libflatccrt library. See section [Cross-compilation](#cross-compilation) for an example.
+
```
Comment (mikkelfj) [inline README.md:833]: suggest adding section:

Note: normally COMMON should also be specified along with BUILDER unless common files are generated by another build step. For simple configurations ALL is likely the best choice.

File: README.md [hunk 3]
```diff
@@ -2358,33 +2351,39 @@ Docker image:
 
 ### Cross-compilation
 
-Users have been reporting some degree of success using cross compiles
-from Linux x86 host to embedded ARM Linux devices.
+Cross-compilation is possible when using the CMake build system. First you have
+to build and install flatcc for the build architecture. Then when cross-compiling
+provide the install location of the build architecture flatcc package via
+environment variable `flatccCli_ROOT` or via cmake variable `flatccCli_ROOT`.
+
```
Comment (mikkelfj) [inline README.md:2366]: Can we have all uppercase ENV variables here?

File: README.md [hunk 4]
```diff
@@ -2358,33 +2351,39 @@ Docker image:
...
+The tests cannot be run because they are compiled for the host architecture and
+thus cannot run on the build machine, so build with cmake option `FLATCC_TEST` disabled.
+The option `FLATCC_RTONLY` will disable tests and only build the runtime library.
+It is highly recommended to at least run the `tests/monster_test`
+project on a new platform.
```
Comment (mikkelfj) [inline README.md:2371]: > It is highly recommended to at least run the `tests/monster_test`
project on a new platform.

True, but exactly what should the user do to achieve that?

File: README.md (hunk 3 above)
Comment (madebr) [reply README.md:2366]: This is a CMake feature for overriding locations of cmake modules.
See https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure

These should be changed to `flatcccli_ROOT` btw.

File: README.md (hunk 4 above)
Comment (wdobbe) [reply README.md:2371]: I left that in from the original cross-building section.
User could cross-compile the samples directory. However to do so some he/she has to make some changes to the CMakeLists.txt in the samples directory.

File: README.md (hunk 2 above)
Comment (wdobbe) [reply README.md:833]: Ok, I will add that fragment.

File: README.md (hunk 1 above)
Comment (wdobbe) [reply README.md]: Ok, I will change that.

File: README.md (hunk 4 above)
Comment (mikkelfj) [reply README.md:2371]: We could also have a cross build example project?

Comment (wdobbe) [reply README.md:2371]: For example, to (cross-)compile only the monster sample, you have to insert following lines on top of the corresponding CMakeLists.txt:

```cmake
cmake_minimum_required (VERSION 2.8.12)
project(flatcc_samples_monster)

find_package(flatcc)
include(FlatccGenerateSources)
```

If you want to add that code permanently, it should be something like:

```cmake
cmake_minimum_required (VERSION 2.8.12)
project(flatcc_samples_monster)

get_directory_property(hasParent PARENT_DIRECTORY)
if(NOT hasParent)
    find_package(flatcc)
    include(FlatccGenerateSources)
endif()
```

Comment (mikkelfj) [reply README.md:2371]: I was thinking about adding a directory parallel to monster that builds independently of the main CMake project and which can cross compile and somehow document that sensibly? Not sure what to call it though.

Comment (mikkelfj) [reply README.md:2371]: And perhaps have a driver script so I go enter ubuntu x64 and cross build to arm or fail due to missing tooling? Could even be a docker container. Just thinking out loud.

**Comment by madebr** [comment]
I have created a `cmake_cross` branch with cross building support.
It's testing a mingw cross build (build: Linux, host: Windows)
See https://github.com/madebr/flatcc/runs/1591287587?check_suite_focus=true

**Comment by mikkelfj** [comment]
awesome

**Comment by mikkelfj** [comment]
except the bross build part :)

**Comment by madebr** [comment]
Haha, I just saw that myself.
bros4life :rofl:

File: README.md (hunk 4 above)
Comment (wdobbe) [reply README.md:2371]: Well, that is basically what Conan does....

**Comment by madebr** [comment]
I'm going to rebase this pr, and rewrite the things for cross build a bit.

**Comment by wdobbe** [comment]
Ok, there is problem with the 'Using flatcc in CMake projects' header link in the readme.
Will commit in a minute or so.

**Comment by madebr** [comment]
@wdobbe 
I've added your commits to the cmake_dep branch, with some changes.
Can you review them?

**Comment by mikkelfj** [comment]
I can't really review the cross build steps - as I don't wan to get into that level of detail. But as long as it is limited to an isolated ci build and it builds correctly, and it doesn't take unreasonable time, it is probably fine.

A read the README changes - looked OK.

One concern is adding a cross file for mingw in the cmake directory, because why that and not x other targets? It seems more appropriate in a test directory, if possible.

**Comment by mikkelfj** [comment]
> Ok, there is problem with the 'Using flatcc in CMake projects' header link in the readme.
Will commit in a minute or so.

I'm using some Vim plugin that updates those links automatically.

**Comment by wdobbe** [comment]
> @wdobbe
> I've added your commits to the cmake_dep branch, with some changes.
> Can you review them?

To make cross-compile work for me I had to make the following change:
```
--- a/cmake/FlatccGenerateSources.cmake
+++ b/cmake/FlatccGenerateSources.cmake
@@ -344,6 +344,7 @@ if(NOT TARGET flatcc::cli_native)
         find_program(FLATCC_BIN
             NAMES flatcc flatcc_d
             PATHS "${FLATCC_ROOT_FOR_BUILD}" ENV FLATCC_ROOT_FOR_BUILD
+            NO_DEFAULT_PATH
             PATH_SUFFIXES bin
             DOC "Path of native flatcc executable"
         )
```
Without NO_DEFAULT_PATH cmake still selects the flatcc exe from the CMAKE_PREFIX_PATH i.s.o. from the PATHS hints.

**Comment by wdobbe** [comment]
A small disadvantage of this solution is that cross-compile only succeeds if you use the `flatcc_generate_sources` function, but I can live with that.

**Comment by madebr** [comment]
@wdobbe 
Do you have
```cmake
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
```
set in your toolchain file?

I was also contemplating creating `flatcc::cli_native` inside the `flatcccli-config.cmake` file, using e.g. a `FindFlatCCNative.cmake` call.
Maybe that would also be a good approach.

**Comment by wdobbe** [comment]
Looking at the find_program documentation, the search order without NO_DEFAULT_PATH is:

1. CMAKE_PREFIX_PATH
2. Search the paths specified by the HINTS option.

So the behaviour we noticed is explainable.

**Comment by wdobbe** [comment]
> @wdobbe
> Do you have
> 
> ```cmake
> set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
> ```
> 
> set in your toolchain file?
> 
I didn't, but if I re-enable that line, cmake still selects the host flatcc exe without `NO_DEFAULT_PATH`.

**Comment by wdobbe** [comment]
> I was also contemplating creating `flatcc::cli_native` inside the `flatcccli-config.cmake` file, using e.g. a `FindFlatCCNative.cmake` call.
> Maybe that would also be a good approach.

As long as it doesn't make the cmake files more complicated than they already are. I think it is acceptable if cross-compiling only works with the `flatcc_generate_sources` function, but we should add that in the README.md cross-compilation section.

**Comment by madebr** [comment]
You need to remove your cache, or at least remove `FLATCC_BIN` from your cache.

**Comment by wdobbe** [comment]
> You need to remove your cache, or at least remove `FLATCC_BIN` from your cache.

I did an `rm -rf *` before running cmake.

**Comment by madebr** [comment]
Ok, maybe adding `NO_DEFAULT_PATH` is the safe way to avoid picking up a flatcc from another prefix.

**Comment by mikkelfj** [comment]
> rm -rf * 

That's how Toy Story 2 got deleted.

**Comment by mikkelfj** [comment]
https://thenextweb.com/media/2012/05/21/how-pixars-toy-story-2-was-deleted-twice-once-by-technology-and-again-for-its-own-good/

**Comment by wdobbe** [comment]
> > rm -rf *
> 
> That's how Toy Story 2 got deleted.

LOL :-) .
Once a year or so I type `rm -rf *` in the wrong terminal. But that's why we have Git.

**Comment by wdobbe** [comment]
> Ok, maybe adding `NO_DEFAULT_PATH` is the safe way to avoid picking up a flatcc from another prefix.

I think we have to. The reason it selects the host exe is because I specified the location of the flatcc host installation dir with `CMAKE_PREFIX_PATH` (which is the normal way with cmake).
And the paths in `CMAKE_PREFIX_PATH` are higher in the search order of find_program that the paths provided with the HINTS keyword.

**Comment by madebr** [comment]
Indeed.
Thinking about it, flatcc, flatccruntime and flatcccli can be merged now.


working now: https://github.com/madebr/flatcc/runs/1592459839?check_suite_focus=true

**Pull request closed without merging**